server/src/main/java/org/opensearch/transport/InboundMessage.java

```diff
@@ -32,105 +32,77 @@
-    private StreamInput streamInput;
+    private final NativeInboundMessage nativeInboundMessage;
 
     public InboundMessage(Header header, ReleasableBytesReference content, Releasable breakerRelease) {
-        this.header = header;
-        this.content = content;
-        this.breakerRelease = breakerRelease;
-        this.exception = null;
-        this.isPing = false;
+        this.nativeInboundMessage = new NativeInboundMessage(header, content, breakerRelease);
     }
 
     public InboundMessage(Header header, Exception exception) {
-        this.header = header;
-        this.content = null;
-        this.breakerRelease = null;
-        this.exception = exception;
-        this.isPing = false;
+        this.nativeInboundMessage = new NativeInboundMessage(header, exception);
     }
 
     public InboundMessage(Header header, boolean isPing) {
-        this.header = header;
-        this.content = null;
-        this.breakerRelease = null;
-        this.exception = null;
-        this.isPing = isPing;
+        this.nativeInboundMessage = new NativeInboundMessage(header, isPing);
     }
 
     public Header getHeader() {
-        return header;
+        return this.nativeInboundMessage.getHeader();
     }
 
     public int getContentLength() {
-        if (content == null) {
-            return 0;
-        } else {
-            return content.length();
-        }
+        return this.nativeInboundMessage.getContentLength();
     }
 
     public Exception getException() {
-        return exception;
+        return this.nativeInboundMessage.getException();
     }
 
     public boolean isPing() {
-        return isPing;
+        return this.nativeInboundMessage.isPing();
     }
 
     public boolean isShortCircuit() {
-        return exception != null;
+        return this.nativeInboundMessage.getException() != null;
     }
 
     public Releasable takeBreakerReleaseControl() {
-        final Releasable toReturn = breakerRelease;
-        breakerRelease = null;
-        if (toReturn != null) {
-            return toReturn;
-        } else {
-            return () -> {};
-        }
+        return this.nativeInboundMessage.takeBreakerReleaseControl();
     }
 
     public StreamInput openOrGetStreamInput() throws IOException {
-        assert isPing == false && content != null;
-        if (streamInput == null) {
-            streamInput = content.streamInput();
-            streamInput.setVersion(header.getVersion());
-        }
-        return streamInput;
+        return this.nativeInboundMessage.openOrGetStreamInput();
     }
 
     @Override
     public void close() {
-        IOUtils.closeWhileHandlingException(streamInput);
-        Releasables.closeWhileHandlingException(content, breakerRelease);
+        this.nativeInboundMessage.close();
     }
 
     @Override
     public String toString() {
-        return "InboundMessage{" + header + "}";
+        return this.nativeInboundMessage.toString();
     }
+
+    @Override
```

Comment:
So since we removed default method, it will impact any code that subclasses `InboundMessage`:

```
class MyInboundMessage extends InboundMessage {
}
```

won't compile anymore.

Reply:
We are overriding the `getProtocol` in `InboundMessage` though. The subclass extending it would not need to implement it, would it? Am I missing something?

Reply:
Yes, certainly, I was overthinking the problem, should be fine, thank you!